src/relay/pass/to_cps.cc

```diff
@@ -0,0 +1,397 @@
+        m->Add(cps_gv, ToCPS(m->Lookup(gv), m, cm));
+      }
+      return k(cm->at(gv));
+    }
+
+    Expr VisitExpr_(const RefCreateNode* op, const MCont& k) final {
+      return VisitExpr(op->value, [&](const Expr& v) { return k(RefCreateNode::make(v)); });
+    }
+
+    Expr reify(const MCont& k) {
+      Var arg = VarNode::make("arg", Type());
+      return FunctionNode::make({arg}, k(arg), Type(), {}, {});
+    }
+
+    Expr reify(const MCont& k, const std::function<Expr(MCont)>& cont) {
+      return LetList::Let(reify(k),
+                          [&](const Var& f) {
+        return cont([&](const Expr& e) { return CallNode::make(f, {e}); });
+      });
+    }
+
+    Expr VisitExpr_(const IfNode* op, const MCont& k) final {
+      return reify(k, [&](const MCont& kf) {
+        return VisitExpr(op->cond,
+                         [&](const Expr& v) {
+          return IfNode::make(v, VisitExpr(op->true_branch, kf), VisitExpr(op->false_branch, kf));
+        });
+      });
+    }
+
+    Expr VisitExpr_(const MatchNode* op, const MCont& k) final {
+      return reify(k, [&](const MCont& kf) {
+        return VisitExpr(op->data, [&](const Expr& v) {
+          tvm::Array<Clause> clauses;
+          for (const auto& c : op->clauses) {
+            clauses.push_back(ClauseNode::make(VisitPattern(c->lhs), VisitExpr(c->rhs, kf)));
+          }
+          return MatchNode::make(v, clauses);
+        });
+      });
+    }
+
+    Expr VisitExpr_(const RefReadNode* op, const MCont& k) final {
+      return VisitExpr(op->ref,
+                       [&](const Expr& r) {
+        return LetList::Let(RefReadNode::make(r), k);
+      });
+    }
+
+    Expr VisitExpr_(const RefWriteNode* op, const MCont& k) final {
+      return VisitExpr(op->ref,
+                       [&](const Expr& r) {
+        return VisitExpr(op->value,
+                         [&](const Expr& v) {
+          return LetList::Let(RefWriteNode::make(r, v), k);
+        });
+      });
+    }
+
+    Expr VisitExpr_(const TupleNode* op, const MCont& k) final {
+      tvm::Array<Expr> fields;
+      std::function<Expr()> next;
+      next = [&]() {
+        return (fields.size() == op->fields.size()) ?
+          k(TupleNode::make(fields)) :
+          VisitExpr(op->fields[fields.size()], [&](const Expr& v) {
+            fields.push_back(v);
+            return next();
+          });
+      };
+      return next();
+    }
+
+    Expr VisitExpr_(const TupleGetItemNode* op, const MCont& k) final {
+      return VisitExpr(op->tuple, [&](const Expr& v) {
+        return k(TupleGetItemNode::make(v, op->index));
+      });
+    }
+
+    Expr VisitExpr_(const CallNode* op, const MCont& k) final {
+      if (op->op.as<OpNode>() || op->op.as<ConstructorNode>()) {
+        tvm::Array<Expr> args;
+        std::function<Expr()> next;
+        next = [&]() {
+          if (args.size() == op->args.size()) {
+            return LetList::Let(CallNode::make(op->op, args), k);
```

Comment:
`CallNode::make(op->op, args, op->attrs, op->type_args)` otherwise you lose attributes